source/string.c

```diff
@@ -34,7 +34,10 @@ struct aws_string *aws_string_new_from_array(struct aws_allocator *allocator, co
 }
 
 struct aws_string *aws_string_new_from_string(struct aws_allocator *allocator, const struct aws_string *str) {
-    return aws_string_new_from_array(allocator, str->bytes, str->len);
+    if (str->allocator) return aws_string_new_from_array(allocator, str->bytes, str->len);
```

Comment:
Hmm... Would this be surprising behavior for a caller? We should clearly document it at a minimum.